wgengine/magicsock/pathfinder.go

```diff
@@ -4,10 +4,101 @@
 
 package magicsock
 
-// startPathFinder initializes the atomicSendFunc, and
+import (
+	"fmt"
+	"net/netip"
+	"time"
+
+	"tailscale.com/tstime/mono"
+	"tailscale.com/types/key"
+)
+
+// startPathFinder initializes the sendFunc, and
 // will eventually kick off a goroutine that monitors whether
 // that sendFunc is still the best option for the endpoint
 // to use and adjusts accordingly.
 func (de *endpoint) startPathFinder() {
-	de.pathFinderRunning = true
+	go func() {
+		for mono.Since(de.lastSendAtomic.Load()) < sessionActiveTimeout {
+			de.pathFinderRunning = true
+
+			// while the session has not timed out yet,
+			// check whether path needs to be upgraded on an interval
+			de.updateSendPathIfNecessary(mono.Now())
+
+			// TODO(2022-10-20): should not be using heartbeat at all, currently just
+			// trying to replicate existing behaviour
+			time.Sleep(heartbeatInterval)
+		}
+	}()
+}
+
+// updateSendPathIfNecessary optionally upates sendFunc
+// based on analysis of current conditions
+func (de *endpoint) updateSendPathIfNecessary(now mono.Time) {
+	de.mu.Lock()
+	defer de.mu.Unlock()
+
+	// default happy state is: use UDP, don't use Derp
+	useUDP := true
+	useDerp := false
+
+	if !de.canP2P() {
+		// if it can't P2P, stick to using only derp
+		useDerp = true
+		useUDP = false
+	}
+
+	// if it's been longer than 6.5 seconds, also send useDerp
+	if now.After(de.trustBestAddrUntil) {
+		useDerp = true
+	}
+
+	derpAddr := de.derpAddr
+	udpAddr := de.bestAddr.AddrPort
+
+	// do final checks to make sure the addresses we want to send to are valid
+	if useUDP && !udpAddr.IsValid() {
+		de.c.logf(fmt.Sprintf("magicsock: silent-disco: invalid UDP addr found: %s", udpAddr))
+		return
+	}
+	if useDerp && !derpAddr.IsValid() {
+		de.c.logf(fmt.Sprintf("magicsock: silent-disco: invalid DERP addr found: %s", derpAddr))
+		return
+	}
+
+	if useUDP && useDerp {
+		de.sendFunc.Store(de.sendDerpAndUDP(udpAddr, derpAddr, de.publicKey))
+	} else if useUDP {
+		de.sendSinglePath(udpAddr, de.publicKey)
+	} else if useDerp {
+		de.sendSinglePath(derpAddr, de.publicKey)
+	}
```

Comment:
these return a func and then you're doing nothing with the result. Did you mean to Store it in the sendFunc?

Reply:
I did, fixed.